Generics.md

```diff
@@ -1,164 +1,386 @@
+
+```rs
+@const @must_use fn array<T, N>(e1 : T, ..., eN : T) -> array<T, N>
+@must_use fn arrayLength(p: ptr<storage, array<E>, AM>) -> u32
+```
+`array` is parameterized by `T` and `N`. `arrayLength` is parameterized by `E` and `AM`.
+* `T` and `E` must be one of: a `scalar`, `vector`, `matrix`, `atomic`, `array`, structure[^1].
+* `N` must be an override-expression that evaluates to a `i32` or `u32`, and greater than 0[^2].
+* `AM` must be a value of enumerant [`access mode`](https://www.w3.org/TR/WGSL/#access-mode)[^3].
+
+So obviously, this level of constraining is too high to express in the grammar.
+
+TL;DR instead, we could specify these constraints:
+* Unconstrained (or implicitly constrained by [AllTypes](https://www.w3.org/TR/WGSL/#alltypes-type).
+* Constrained by: a user-defined struct, `vec`, `mat`, `array`, `atomic`, `ptr`, `AbstractInt` or `AbstractFloat`: Must be specialized by a *constructible* subtype.
+* Constrained by a leaf type: `u32`, `i32`, `f32`, `f16`, `bool`: Must be specialized with a instance of that type (like the `N` parameter).
+
+## Generic Declarations
+
+We introduce a `@generic` attribute to define the generic parameters and constraints.
+```rs
+@generic(Num: AbstractFloat, T: MyStruct, N: u32)
+fn my_fn(x: Num, t: T) -> array<Num, N>
+
+@generic(T)
+struct Point2D {
+  x: T,
+  y: T,
+}
+
+@generic(T: AbstractInt, W: u32, H: u32)
+alias matrix = array<array<T, H>, W>;
+```
+
+### Using generic declarations
+
+I chose the template list notation, as it seems to be unambiguous in the grammar and is a familiar notation for generics. Besides, it is already used for type generators.
+
+```rs
+let res: array<f32, 10> = my_fn<f32, SomeStruct, 10>(3.5f, struct_inst);
+
+@generic(T: AbstractFloat)
+fn rotate(point: Point2D<T>, angle: f32) -> Point2D<T> {}
 
+alias mat4x4f = matrix<f32, 4, 4>;
+```
 
-Simple Example:
+## (optional) Function-as-Type
 
-* ```
-  ./util.wgsl:
+Since a function signature now has a well-defined type and subtyping, we *could* support "function pointers" via genericity.
 
-  @export fn workgroupMin<E>(elems: array<E, 4>) -> E { }  // E is a generic parameter
+You cannot pass by value a function pointer, instead it is passed as a specialization of the generic parameter.
+
+This can be implemented simply by substitution of the generic type, like the rest of the generics. No extra work needed except for the type-checking.
+
+```rs
+@generic(T)
+fn callback_signature(arg: T) -> f32; // just define an abstract function, this serves only the purpose of expliciting the signature.
+
+@generic(T, F: callback_signature<T>)
+fn my_fn() -> f32 {
+  // ...
+  return F(val2); // call the generic function (similar to a function pointer).
+}
+
+fn callback_implementor(arg: u32) -> f32 { // this function signature is a subtype of callback_signature
+  // ...
+}
+
+fn main() {
+  my_fn<u32, callback_implementor>();
+}
+```
+
+See [the map reduce example](#map-reduce).
+
+## Generic Modules
```

Comment:
Is it clear that generic modules are the best way forward to add ergonmics to generic elements? If so, let's discuss here, if it's more debatable, perhaps we should split this part into a separate issue/PR.